extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/exec/LoadedSegmentDataProviderTest.java

```diff
@@ -0,0 +1,199 @@
+        .intervals(new MultipleIntervalSegmentSpec(ImmutableList.of(Intervals.of("2003/2004"))))
+        .columns("__time", "cnt", "dim1", "dim2", "m1", "m2", "unique_dim1")
+        .resultFormat(ScanQuery.ResultFormat.RESULT_FORMAT_COMPACTED_LIST)
+        .context(ImmutableMap.of(QueryContexts.NUM_RETRIES_ON_MISSING_SEGMENTS_KEY, 1))
+        .build();
+    QueryToolChestWarehouse queryToolChestWarehouse = new MapQueryToolChestWarehouse(
+        ImmutableMap.<Class<? extends Query>, QueryToolChest>builder()
+                    .put(ScanQuery.class, new ScanQueryQueryToolChest(null, null))
+                    .build()
+    );
+    target = spy(
+        new LoadedSegmentDataProvider(
+            DATASOURCE1,
+            new ChannelCounters(),
+            mock(ServiceClientFactory.class),
+            coordinatorClient,
+            DruidServiceTestUtils.newJsonMapper(),
+            queryToolChestWarehouse,
+            Execs.scheduledSingleThreaded("query-cancellation-executor")
+        )
+    );
+    doReturn(dataServerClient).when(target).makeDataServerClient(any());
+  }
+
+  @Test
+  public void testFetchRowsFromServer() throws IOException
+  {
+    doReturn(Sequences.simple(ImmutableList.of(scanResultValue))).when(dataServerClient).run(any(), any(), any(), any());
+
+    Pair<LoadedSegmentDataProvider.DataServerQueryStatus, Yielder<Object[]>> dataServerQueryStatusYielderPair = target.fetchRowsFromDataServer(
+        query,
+        SEGMENT_1,
+        ScanQueryFrameProcessor::mappingFunction,
+        Closer.create()
+    );
+
+    Assert.assertEquals(LoadedSegmentDataProvider.DataServerQueryStatus.SUCCESS, dataServerQueryStatusYielderPair.lhs);
+    List<List<Object>> events = (List<List<Object>>) scanResultValue.getEvents();
+    Yielder<Object[]> yielder = dataServerQueryStatusYielderPair.rhs;
+    events.forEach(
+        event -> {
+          Assert.assertArrayEquals(event.toArray(), yielder.get());
+          yielder.next(null);
+        }
+    );
+  }
+
+  @Test
+  public void testHandoff() throws IOException
+  {
+    doAnswer(invocation -> {
+      ResponseContext responseContext = invocation.getArgument(1);
+      responseContext.addMissingSegments(ImmutableList.of(SEGMENT_1));
+      return Sequences.empty();
+    }).when(dataServerClient).run(any(), any(), any(), any());
+    doReturn(Futures.immediateFuture(Boolean.TRUE)).when(coordinatorClient).isHandoffComplete(DATASOURCE1, SEGMENT_1);
+
+    Pair<LoadedSegmentDataProvider.DataServerQueryStatus, Yielder<Object[]>> dataServerQueryStatusYielderPair = target.fetchRowsFromDataServer(
+        query,
+        SEGMENT_1,
+        ScanQueryFrameProcessor::mappingFunction,
+        Closer.create()
+    );
+
+    Assert.assertEquals(LoadedSegmentDataProvider.DataServerQueryStatus.HANDOFF, dataServerQueryStatusYielderPair.lhs);
+    Assert.assertNull(dataServerQueryStatusYielderPair.rhs);
+  }
+
+
+  @Test
+  public void testQueryFail()
```

Comment:
Can you add a test case for server not found as well ? and check if it reroutes the query to the next server ?

Reply:
Added 2 new tests